### PR TITLE
QOL improvements for -j trace and |mass outputs

### PR DESCRIPTION
**Pull request by ixv**
The output of -j tracer is now split every 100,000 lines so that chrome doesn't choke on large traces (is this a reasonable number?) Also stuck the output of these in the put directory under a subdirectory called trace, and named them by unix date.

Stuck |mass outputs in their own subdirectory of put, called mass, for better organization.

### CODE REVIEW

File: vere/raft.c [hunk 1]
```diff
@@ -1868,7 +1869,14 @@ _raft_grab(u3_noun rus)
...
       c3_c nam_c[2048];
-      snprintf(nam_c, 2048, "%s/.urb/put/%s-mass.txt", u3_Host.dir_c, wen_c);
+      snprintf(nam_c, 2048, "%s/.urb/put/mass", u3_Host.dir_c);
+
+      struct stat st;
+      if ( -1 == stat(nam_c, &st) ) {
+        mkdir(nam_c, 0700);
+      }
+
+      snprintf(nam_c, 2048, "%s/%s.txt", nam_c, wen_c);
```
Comment (joemfb) [inline vere/raft.c]: I think this nested directly structure is fine. Do you think it's worth putting the ship name in the file for ease of later handling?

File: noun/trace.c [hunk 2]
```diff
@@ -300,14 +303,25 @@ u3t_flee(void)
...
+
+  c3_c fil_c[2048];
+  snprintf(fil_c, 2048, "%s/.urb/put/trace", u3_Host.dir_c);
+
+  struct stat st;
+  if ( -1 == stat(fil_c, &st) ) {
+    mkdir(fil_c, 0700);
+  }
+
+  snprintf(fil_c, 2048, "%s/%ld.json", fil_c, time(NULL));
```
Comment (joemfb) [inline noun/trace.c]: I think it'd be better to pretty-print `u3A->wen`, to be consistent with the way |mass effect files are named. Likewise, if we add the ship name there, we should do so here too.

File: vere/raft.c (hunk 1 above)
Comment (ixv) [reply vere/raft.c]: sure that sounds good.

File: noun/trace.c (hunk 2 above)
Comment (ixv) [reply noun/trace.c]: nevermind, I found a solution

Comment (joemfb) [reply noun/trace.c]: Ok, good. You can't use the loom in any way prior to `u3m_boot()`. Exactly when `u3A->wen` or `u3A->own` is set and available depends on whether it's a first boot or restart. `u3_sist_boot()` is generally the save place to put things that depend on arvo state.

Comment (ixv) [reply noun/trace.c]: we probably want the profiler to catch boot events though right?

Comment (joemfb) [reply noun/trace.c]: Yeah, that's probably more important. It'll be much easier to do this right on cc-release.

Comment (ixv) [reply noun/trace.c]: so I can get u3A->wen pretty early in the process, so I'm gonna just name it by that and leave off the ship name. I'll leave off the ship name for |mass as well, for consistency's sake.

**Comment by eglaysher** [comment]
I'm uncertain if this approach actually works, as beginning and end events won't line up. I suspect that something like this is possible if you instead align to event boundaries instead of the number of lines emitted.

File: noun/vortex.c [hunk 3]
```diff
@@ -102,6 +102,8 @@ u3v_start(u3_noun now)
     printf("arvo: time: %s\n", wen_c);
     free(wen_c);
   }
+
+  u3t_trace_open();
```
Comment (joemfb) [inline noun/vortex.c]: Does this need to be inside a `u3C.wag_w & u3o_trace` condition?

Comment (ixv) [reply noun/vortex.c]: yes lol

Comment (ixv) [reply noun/vortex.c]: fixed!

**Comment by ixv** [comment]
Are you saying we should only check if trace_counter >= 100000 inside u3t_event_trace, but not u3t_nock_trace?

**Comment by eglaysher** [comment]
No, I'm saying that we should only cycle trace logs between arvo events. I
don't believe the trace format deals with incomplete events properly.


—
~littel-ponnys
https://urbit.org

**Comment by joemfb** [comment]
Then it seems like it should work to just cycle the files (and increment the counter) in `u3t_event_trace()`.

**Comment by ixv** [comment]
I split the file in u3_raft_chip. Also made a u3_trac struct inside u3_Host with the tracing information so that raft.c has access.

**Pull request merged**